test/sfs.py

```diff
@@ -81,48 +84,197 @@ def open(file_name, mode='r'):
+
+    def writable(self):
+        """Return True if the stream supports writing."""
+        return bool(self._flags & NFS4_WRONLY)
+
+    def writelines(self, lines):
+        """
+        Write a list of lines to the stream.
+
+        Line separators are not added, so it is usual for each of the lines provided to have a line separator at the end.
+        """
+        total_bytes_written = 0
+        for line in lines:
+            bytes_written = self.write(line)
+            if bytes_written < 0:
+                return total_bytes_written
+            total_bytes_written += bytes_written
+        return total_bytes_written
 
     def read(self, size=-1):
+        """
+        Read up to size bytes from the object and return them.
+
+        If size is -1, all bytes until EOF are returned with possibly 
+        multiple underlying calls. If size is specified, only one call will be
+        made.
+
+        If 0 bytes are returned and size was not 0, the end of file has been reached.
+        """
+        if size == -1:
+            return self.readall()
+        elif size < 0:
+            raise ValueError("size must be >= -1")
         buffer = ctypes.create_string_buffer(size)
-        read_status = c_helper.nfs4_pread(self._file, buffer, self._pos, size)
-        if read_status != NFS4_OK:
-            if read_status == NFS4ERR_OPENMODE:
+        bytes_read = c_helper.nfs4_pread(self._file, buffer, self._pos, size)
+        if bytes_read < 0:
+            if c_helper.nfs4_error_num(client) == NFS4ERR_OPENMODE:
                 raise io.UnsupportedOperation("not readable") 
             print("Failed to read file: " + c_helper.nfs4_error_string(client).decode(encoding='utf-8'))
             return
-        self._pos += size
+        self._pos += len(buffer.value)
         return buffer.value
 
+    def readall(self):
+        """
+        Read and return all the bytes from the stream until EOF.
+        
+        Use multiple calls if necessary. Return None upon error.
+        """
+        segments = b''
+
+        segment = self.read(io.DEFAULT_BUFFER_SIZE)
+        while segment:
+            segments += segment
+            segment = self.read(io.DEFAULT_BUFFER_SIZE)
+        return segments
+
+    def readinto(self, b):
+        """
+        Read bytes into a pre-allocated, writable bytes-like object b, and return the number of bytes read. 
+
+        If the object is in non-blocking mode and no bytes are available, None is returned.
+        """
+        length = len(b)
+        data = self.read(length)
+        for i in range(len(data)):
+            b[i] = data[i]
+        return len(data) or None
+
     def write(self, content_bytes):
-        content_len = len(content_bytes)
-        write_status = c_helper.nfs4_pwrite(self._file, content_bytes, self._pos, content_len)
-        if write_status != NFS4_OK:
-            if write_status == NFS4ERR_OPENMODE:
+        array = ctypes.c_byte * len(content_bytes)
```

Comment:
@chenlibo do you know if there's a better way to be doing this with ctypes? Specifically moving content bytes from `bytes` to a `c_void_p` friendly type?

Reply:
Hmm, what if the user wants to write straight random bytes? The `bytes(content_bytes, 'utf-8')'` will probably fail. 

I suppose that's what the `b` flag and binary mode is for with Python `open`. Do you think we should support the b flag so users can pass in straight strings / straight bytes?

Reply:
I think so. That may require a major change in how we implement buffering?